flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/plan/nodes/resource/ExecNodeResourceTest.scala

```diff
@@ -0,0 +1,284 @@
+    val table = testUtil.getTableEnv.sqlQuery(sqlQuery)
+    val tableSink = new MockTableSink(new TableSchema(Array("a", "b", "c"),
+      Array[TypeInformation[_]](Types.INT, Types.LONG, Types.STRING)), 25)
+
+    testUtil.getTableEnv.writeToSink(table, tableSink, "sink")
+    testUtil.getTableEnv.generateStreamGraph()
+    assertEquals(25, tableSink.getSinkTransformation.getParallelism)
+  }
+}
+
+object ExecNodeResourceTest {
+
+  @Parameterized.Parameters(name = "isBatch={0}, {1}")
+  def parameters(): util.Collection[Array[Any]] = {
+    util.Arrays.asList(
+      Array(true, InferMode.NONE),
+      Array(true, InferMode.ONLY_SOURCE),
+      Array(false, InferMode.NONE),
+      Array(false, InferMode.ONLY_SOURCE)
+    )
+  }
+
+  def setResourceConfig(tableConfig: TableConfig): Unit = {
+    tableConfig.getConf.setInteger(
+      TableConfigOptions.SQL_RESOURCE_DEFAULT_PARALLELISM,
+      18)
+    tableConfig.getConf.setInteger(
+      TableConfigOptions.SQL_RESOURCE_INFER_SOURCE_PARALLELISM_MAX,
+      1000)
+    tableConfig.getConf.setLong(
+      TableConfigOptions.SQL_RESOURCE_INFER_ROWS_PER_PARTITION,
+      1000000
+    )
+  }
+}
+
+/**
+  * Batch/Stream [[org.apache.flink.table.sources.TableSource]] for resource testing.
+  */
+class MockTableSource(isBatch: Boolean, schema: TableSchema)
+    extends StreamTableSource[BaseRow] {
+
+  override def isBounded: Boolean = isBatch
+
+  override def getDataStream(
+      execEnv: environment.StreamExecutionEnvironment): DataStream[BaseRow] = {
+    val transformation = mock(classOf[SourceTransformation[BaseRow]])
+    when(transformation.getMaxParallelism).thenReturn(-1)
+    val bs = mock(classOf[DataStream[BaseRow]])
+    when(bs.getTransformation).thenReturn(transformation)
+    when(transformation.getOutputType).thenReturn(getReturnType)
+    val factory = mock(classOf[StreamOperatorFactory[BaseRow]])
+    when(factory.isStreamSource).thenReturn(!isBatch)
+    when(transformation.getOperatorFactory).thenReturn(factory)
+    bs
+  }
+
+  override def getReturnType: TypeInformation[BaseRow] = {
+    val LogicalTypes = schema.getFieldTypes.map(
+      TypeInfoLogicalTypeConverter.fromTypeInfoToLogicalType)
+    new BaseRowTypeInfo(LogicalTypes, schema.getFieldNames)
+  }
+
+  override def getTableSchema: TableSchema = schema
+}
+
+/**
+  * Batch/Stream [[org.apache.flink.table.sinks.TableSink]] for resource testing.
+  */
+class MockTableSink(schema: TableSchema, maxParallelism: Int) extends StreamTableSink[BaseRow]
+with AppendStreamTableSink[BaseRow] {
+
+  private var sinkTransformation:SinkTransformation[BaseRow] = _
+
+  override def emitDataStream(dataStream: DataStream[BaseRow]) = ???
+
+  override def consumeDataStream(dataStream: DataStream[BaseRow]): DataStreamSink[_] = {
+    val outputFormat = mock(classOf[OutputFormat[BaseRow]])
+    val ret = dataStream.writeUsingOutputFormat(outputFormat).name("collect")
+    ret.getTransformation.setParallelism(17)
```

Comment:
minor: could we make the number as a constructor parameter or make it to be a public static final variable?

Reply:
Because the TableSink is only used in this ut, so there is necessary that it is changed to a variable.

Reply:
The reason I comment above is I find the test is asserting 17 parallelism in `testSinkConfigParallelismWhenMax1`. However, I didn't find a clue of 17 in the test until I looked into the `MockTableSink`. So I think it is a bit obscure. But it's not a big problem.